File: product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:

    // O(n^2) solution
    // O(1) space
    // vector<int> productExceptSelf(vector<int>& nums) {

    //     vector<int> sol; 

    //     for (int i = 0; i < nums.size(); i++){

    //         int prod = 1; 

    //         for (int j = 0; j < nums.size(); j++){
    //             if (j == i) continue; 
    //             prod *= nums[j];
    //         }
    //         sol.push_back(prod); 
    //     }

    //     return sol; 
    // }


    vector<int> productExceptSelf(vector<int>& nums) {

        vector<int> ans(nums.size(), 1);

        // Stores all values to the left of nums[i]
        // We get this by taking the value to the left of nums[i], and multiple it by 
        // All the vlaues multipled to the left of it, which is stored in ans
        for (int i = 1; i < nums.size(); i++){
            ans[i] = nums[i - 1] * ans[i - 1];
        }

        int right = 1; 

        for (int i = nums.size() - 1; i >= 0; i--){
            ans[i] = ans[i] * right; 
            right *= nums[i];
        }

        return ans; 
    }


};

File: product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        # Start getting the left prefix product
        ans = [1]

        for i in range(1, len(nums)):
            ans.append(ans[i - 1] * nums[i - 1])
    
        postfix = 1

        for i in range(len(nums) - 1, -1, -1):
            ans[i] *= postfix
            postfix *= nums[i]

        return ans

